cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/TimePartitionedStreamFileWriter.java

```diff
@@ -152,17 +157,37 @@ public int hashCode() {
 
       // Create the event and index file with the next sequence id
       int fileSequence = StreamUtils.getNextSequenceId(partitionDirectory, fileNamePrefix);
-      Location eventFile = StreamUtils.createStreamLocation(partitionDirectory, fileNamePrefix,
-                                                            fileSequence, StreamFileType.EVENT);
-      Location indexFile = StreamUtils.createStreamLocation(partitionDirectory, fileNamePrefix,
-                                                            fileSequence, StreamFileType.INDEX);
+      final Location eventFile = StreamUtils.createStreamLocation(partitionDirectory, fileNamePrefix,
+                                                                  fileSequence, StreamFileType.EVENT);
+      final Location indexFile = StreamUtils.createStreamLocation(partitionDirectory, fileNamePrefix,
+                                                                  fileSequence, StreamFileType.INDEX);
       // The creation should succeed, as it's expected to only have one process running per fileNamePrefix.
       if (!eventFile.createNew() || !indexFile.createNew()) {
         throw new IOException("Failed to create new file at " + eventFile + " and " + indexFile);
       }
 
       LOG.debug("New stream file created at {}", eventFile);
-      return new StreamDataFileWriter(createOutputSupplier(eventFile), createOutputSupplier(indexFile), indexInterval);
+
+      final StreamDataFileWriter streamDataFileWriter =
+        new StreamDataFileWriter(createOutputSupplier(eventFile), createOutputSupplier(indexFile), indexInterval);
+
+      timer.schedule(new TimerTask() {
```

Comment:
That seems a bit difficult to do because at the `ConcurrentStreamWriter` level, the FileWriter is not guaranteed to be a `PartitionedFileWriter` nor a `TimePartitionedStreamFileWriter`, so we can't really check which is the smallest partition time.

Also, since its operating on the `FileWriter` interface, seems like we need to add this `refresh` method to the `FileWriter` interface.

Alternatively, maybe we can share a single thread pool across all the streams by injecting the executor as singleton via Guice to all the different `TimePartitionedStreamFileWriter`s.